k8s/envoy-jwt-opa/README.md

```diff
@@ -0,0 +1,336 @@
+
+```console
+$ bash scripts/pre-set-env.sh
+```
+
+The script will create all the resources needed for the SPIRE Server and SPIRE Agent to be available in the cluster and then will create all the resources for the SPIRE Envoy JWT tutorial, which is the base scenario for this SPIRE Envoy JWT with OPA Tutorial.
+
+**Note:** The configuration changes needed to enable Envoy and OPA to work with SPIRE are shown as snippets in this tutorial. However, all of these settings have already been configured. You don't have to edit any configuration files.
+
+
+# Part 1: Deploy Updated and New Resources
+
+Assuming the SPIRE Envoy JWT Tutorial as a starting point, there are some resources that need to be created.
+The goal is to have the requests authorized by the OPA Agent before hitting the `backend` service. There is an mTLS connection established between Envoy instances where JWT SVIDs are transmitted in requests as `authorization` headers. So the missing part is to add an OPA Agent to authorize requests based on policies.
+The solution applied in this tutorial consists of adding a new external authorization filter to the Envoy instance running in front of the `backend` service. The new filter invokes the OPA Agent after the request pass through the Envoy JWT Auth Helper (first filter) and its job is to check whether the request should be authorized or denied.
+
+## Update Deployments
+
+In order to let OPA authorize or reject requests coming to the `backend` service it is necessary to add OPA as a sidecar to the deployment.
+We use the `openpolicyagent/opa:latest-istio` image which extends OPA with a gRPC server that implements the Envoy External Authorization API so OPA can communicate policy decisions with Envoy. The new container is added and configured as follow in [`backend-deployment.yaml`](k8s/backend/backend-deployment.yaml):
+
+
+```console
+- name: opa
+  image: openpolicyagent/opa:latest-istio
+  imagePullPolicy: IfNotPresent
+  ports:
+    - name: opa-envoy
+      containerPort: 8182
+      protocol: TCP
+    - name: opa-api-port
+      containerPort: 8181
+      protocol: TCP
+   args:
+     - "run"
+     - "--server"
+     - "--config-file=/run/opa/opa-config.yaml"
+     - "/run/opa/opa-policy.rego"
+   volumeMounts:
+     - name: backend-opa-policy
+       mountPath: /run/opa
+       readOnly: true
+```
+
+The ConfigMap `backend-opa-policy` needs to be added into the `volumeMounts` section, like this:
+
+```console
+- name: backend-opa-policy
+  configMap:
+    name: backend-opa-policy-config
+```
+
+The ConfigMap `backend-opa-policy` provides two resources, `opa-config.yaml` described in [OPA Configuration](#opa-configuration) and the `opa-policy.rego` policy explained in the [Rego Policy](#opa-policy) section.
```

Comment:
If you think the section should be named Rego Policy instead of OPA Policy, feel free to change the section name on line 107 instead of these cross references.

```suggestion
The ConfigMap `backend-opa-policy` provides two resources, `opa-config.yaml` described in [OPA Configuration](#opa-configuration) and the `opa-policy.rego` policy explained in the [OPA Policy](#opa-policy) section.
```